LDAP: No longer shows incorrectly matching groups based on role in debug page

xlson opened the pull request:
**What this PR does / why we need it**:

No longer show all groups that gives a specific Org Role for a user that gets that role from one of them. 

**Which issue(s) this PR fixes**:

<!--

* Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #20016

**Special notes for your reviewer**:

xlson committed:
LDAP Debug: No longer shows incorrectly matching groups based on role

Org Role was used as a shortcut to figure out what groups were matching
and which weren't. That lead to too all groups matching a specific role
to show up for a user if that user got that role.

torkelo reviewed (commented):
I think we can simplify the code a bit by combining the to for loops:

So instead of looping over config groups then user groups (nested), and then user groups again. Should be the same as below: 

```go

	orgRoles := []LDAPRoleDTO{}

	for _, userGroup := range user.Groups {
		matches := false

		for _, configGroup := range serverConfig.Groups {
			if configGroup.GroupDN == userGroup {
				r := &LDAPRoleDTO{GroupDN: configGroup.GroupDN, OrgId: configGroup.OrgId, OrgRole: configGroup.OrgRole}
				orgRoles = append(orgRoles, *r)
				matches = true
				break
			}
		}

		if !matches {
			r := &LDAPRoleDTO{GroupDN: userGroup}
			orgRoles = append(orgRoles, *r)
		}
	}
```

gotjosh commented on pkg/api/ldap_debug.go:
```diff
@@ -278,13 +278,15 @@ func (server *HTTPServer) GetUserFromLDAP(c *models.ReqContext) Response {
 
 	orgRoles := []LDAPRoleDTO{}
 
-	// First, let's find the groupDN that we did match by inspecting the assigned user OrgRoles.
```
I think we need to match all three conditions - there is a world where you can match the group but not the org/role.

The LDAP code is structured in a way that if you have a role assigned for an org it won't match any more beyond the first. The order of how the roles are specified in the TOML file matter.

xlson replied:
Thanks for pointing this out, I wasn't aware of that. I've updated the code to reflect this.

xlson committed:
LDAP Debug: Fixes ordering of matches

The order of groups in the ldap.toml file is important, only the first
match for an organisation will be used. This means we have to iterate
based on the config and stop matching when a match is found.

We might want to think about showing further matches as potential
matches that are shadowed by the first match. That would possibly make
it easier to understand why one match is used instead of another one.

xlson commented:
@torkelo As we match groups based on the order of configs in the toml file I don't think we can do it quite this cleanly (thanks for pointing this out @gotjosh). It seems we need to match based on the toml config as only the first matching role for an org will be used. (We could consider choosing to show the others as matching as well if we have a way to signify that they are being shadowed by the first match). If we choose to iterate based on the user's groups we wouldn't know what the first match was.

I've rewritten the code in a way that seems to match what the real LDAP mapping code is doing. It doesn't look great though as it ends up with a for -> if -> for cycle. I'll see if it's possible to clean that up a bit tomorrow.

xlson commented on pkg/api/ldap_debug_test.go:
```diff
@@ -240,6 +245,7 @@ func TestGetUserFromLDAPApiEndpoint_WithTeamHandler(t *testing.T) {
 		Name:           "John Doe",
 		Email:          "john.doe@example.com",
 		Login:          "johndoe",
+		Groups:         []string{"cn=admins,ou=groups,dc=grafana,dc=org"},
```
@gotjosh I didn't quite understand what we were testing here.

gotjosh replied:
That the team handler is present - which is a command only available on enterprise mode.  It's a very subtle detail: the `teams` key on the payload is `[]` instead of `null`

gotjosh replied:
I could have asserted against _only_ that key but that wouldn't cover changes made to the payload as a whole and I would rather have that coverage.

xlson replied:
Thanks for clarifying.

torkelo commented:
Regarding the fact that only the first match is used (per org). Not sure what is best here. Not showing "Match" and listing them as "No Match" feels wrong as well as it might give the impression that the group is specified wrongly in the ldap toml. As long as we list them in the same way as they mappings are specified in the toml file so that the used match & role is first then maybe that is best? 

We should probably change so the order or mappings is not used and that the highest role/permission instead wins. (But not in this PR).

xlson commented:
I agree, showing as `No Match` is very confusing. Yeah, that might be for the best. We could add an info icon with an explain that only the first match is used.

I agree. While the it is working now makes it possible to create more creative configuration, I think it's likely that we're paying too much for that power by making our code more complicated and the LDAP config harder to understand.

xlson committed:
LDAP Debug: never display more than one match for the same LDAP group/mapping.

xlson committed:
LDAP Debug: show all matches, even if they aren't used

torkelo reviewed (approved):
Looks good

xlson committed:
Update public/app/features/admin/ldap/LdapUserGroups.tsx

Co-Authored-By: gotjosh <josue.abreu@gmail.com>

xlson committed:
Update public/app/features/admin/ldap/LdapUserGroups.tsx

Co-Authored-By: gotjosh <josue.abreu@gmail.com>
